server/src/core/stat/stat-service.ts

```diff
@@ -0,0 +1,108 @@
+import { Db, Collection } from 'mongodb';
+import { createDbOpFailedError, rethrowIfAppError, createResourceNotFoundError } from '../error';
+import { generateId } from '../util';
+import * as messages from '../messages';
+import { UserService }  from '../user';
+import { TransactionService } from '../payment';
+import { StatsService, StatsArgs, Stats, StatsCreateArgs, StatsUpdateArgs } from './types';
+
+const COLLECTION = 'stats';
+export class Statistics implements StatsService {
+  private db: Db;
+  private collection: Collection<Stats>;
+  private users: UserService;
+  private transactions: TransactionService;
+
+  constructor(db: Db, args: StatsArgs) {
+    this.db = db;
+    this.collection = this.db.collection(COLLECTION);
+    this.users = args.users;
+    this.transactions = args.transactions;
+    this.create();
+  }
+
+  private async create(): Promise<Stats> {
+    const stats = await this.get();
+    if (!stats) {
+      const now = new Date();
+      const stat: Stats = {
+        _id: 'stats',
+        numContributors: 0,
+        totalContributed: 0,
+        numBeneficiaries: 0,
+        totalDistributed: 0,
+        updatedAt: now
+      };
+
+      try {
+        const res = await this.collection.insertOne(stat);
+        return res.ops[0];
+      }
+      catch (e) {
+        rethrowIfAppError(e);
+        throw createDbOpFailedError(e.message);
+      }
+    }
+    else return stats;
+  }
+
+  async get(): Promise<Stats> {
+    try {
+      const stats = await this.collection.findOne({ _id: 'stats' });
+      return stats;
+    }
+    catch (e) {
+      rethrowIfAppError(e);
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  async update(): Promise<Stats> {
+    try {
+      const numContributorsRes = await this.users.aggregate([
+        { $match: { roles: { $in: ['donor'] } } },
+        { $count: 'numContributors' }
+      ]);
+
+      const numBeneficiariesRes = await this.users.aggregate([
+        { $match: { roles: { $in: ['beneficiary'] } } },
+        { $count: 'numBeneficiaries' }
+      ]);
```

Comment:
The number of contributors should be the number of people who have made a successful donation (not the number of users who are marked as donors)

Similarly, the number of beneficiaries should be the number of users who have received a donation.

This means that you need to aggregate and filter transactions to find this information. No need to aggregate the users collection.

e.g., for contributors: filter successful donation transactions, get the distinct senders (i.e. the `from` field) and count them.

If it's possible to get the total amount contributed and the number of contributors in the same aggregation, that would be great.

If it's also possible to get the total amount distributed and beneficiaries in one aggregation, that'd also be great.